pkg/stash/with_etcd.go

```diff
@@ -0,0 +1,95 @@
+package stash
+
+import (
+	"context"
+	"time"
+
+	"github.com/gomods/athens/pkg/config"
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/clientv3/concurrency"
+	"golang.org/x/sync/errgroup"
+)
+
+// WithEtcd returns a distributed singleflight
+// using an etcd cluster. If it cannot connect,
+// then it will panic
+func WithEtcd(endpoints []string) (Wrapper, error) {
+	const op errors.Op = "stash.WithEtcd"
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	c, err := clientv3.New(clientv3.Config{
+		Endpoints:   endpoints,
+		DialTimeout: time.Second * 5,
+	})
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	var eg errgroup.Group
+	for _, ep := range endpoints {
+		eg.Go(func() error {
+			_, err := c.Status(ctx, ep)
+			return err
+		})
+	}
+	err = eg.Wait()
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	return func(s Stasher) Stasher {
+		return &etcd{c, s}
+	}, nil
+}
+
+type etcd struct {
+	c *clientv3.Client
+	s Stasher
+}
+
+func (s *etcd) Stash(ctx context.Context, mod, ver string) error {
+	const op errors.Op = "etcd.Stash"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+
+	mv := config.FmtModVer(mod, ver)
+	sesh, err := concurrency.NewSession(s.c)
+	if err != nil {
+		return errors.E(op, err)
+	}
+	mu := concurrency.NewMutex(sesh, mv)
+	err = mu.Lock(ctx)
+	if err != nil {
+		return errors.E(op, err)
+	}
+	defer mu.Unlock(ctx)
+	resp, err := s.c.Get(ctx, mv)
+	if err != nil {
+		return errors.E(op, err)
+	}
+	if resp.Count == 0 {
+		lease, err := s.c.Grant(ctx, 15*60)
+		if err != nil {
+			return errors.E(op, err)
+		}
+		err = s.s.Stash(ctx, mod, ver)
+		stashResponse := "success"
+		if err != nil {
+			stashResponse := err.Error()
+			s.c.Put(ctx, mv, stashResponse, clientv3.WithLease(lease.ID)) // TODO: handle err
+			return err
+		}
+
+		_, err = s.c.Put(ctx, mv, stashResponse, clientv3.WithLease(lease.ID))
```

Comment:
Could it be writen in shortened form 
`if _, err := s.c.Put(ctx, mv, "success", clientv3.WithLease(lease.ID)); err != nil {...`? 
Also I think that `stashResponse` variable seems unnecessary